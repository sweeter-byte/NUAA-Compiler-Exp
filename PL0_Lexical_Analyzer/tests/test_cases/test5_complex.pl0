{test5_complex.pl0 - Complex Comprehensive Test}
{Test: Complete PL/0 program, including all syntax elements}

program complex;
const
    max := 100,
    min := 1,
    step := 2;
var x, y, z, sum, count;

procedure init;
begin
    x := 0;
    y := 0;
    sum := 0;
    count := 0
end;

procedure calculate;
var temp;
begin
    temp := x + y;
    sum := sum + temp;
    count := count + 1
end;

begin
    call init;
    
    (* Main loop: Calculate the sum from 1 to max *)
    x := min;
    while x < max do
    begin
        if odd x then
        begin
            sum := sum + x;
            write(sum)
        end;
        x := x + step
    end;
    
    (* Read user input *)
    read(y, z);
    
    (* Conditional judgment and procedure call *)
    if y > z then
    begin
        x := y;
        y := z;
        call calculate
    end
    else
    begin
        x := z;
        call calculate
    end;
    
    (* Output final result *)
    write(sum, count)
end