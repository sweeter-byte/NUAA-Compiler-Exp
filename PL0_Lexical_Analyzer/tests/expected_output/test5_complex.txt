================================================================================
PL/0 Lexical Analysis Results
================================================================================

Index  Line   Column Type            Value               
--------------------------------------------------------------------------------
1      4      1      PROGRAM         program             
2      4      9      ID              complex             
3      4      16     SEMICOLON       ;                   
4      5      1      CONST           const               
5      6      5      ID              max                 
6      6      9      ASSIGN          :=                  
7      6      12     INT             100                 
8      6      15     COMMA           ,                   
9      7      5      ID              min                 
10     7      9      ASSIGN          :=                  
11     7      12     INT             1                   
12     7      13     COMMA           ,                   
13     8      5      ID              step                
14     8      10     ASSIGN          :=                  
15     8      13     INT             2                   
16     8      14     SEMICOLON       ;                   
17     9      1      VAR             var                 
18     9      5      ID              x                   
19     9      6      COMMA           ,                   
20     9      8      ID              y                   
21     9      9      COMMA           ,                   
22     9      11     ID              z                   
23     9      12     COMMA           ,                   
24     9      14     ID              sum                 
25     9      17     COMMA           ,                   
26     9      19     ID              count               
27     9      24     SEMICOLON       ;                   
28     11     1      PROCEDURE       procedure           
29     11     11     ID              init                
30     11     15     SEMICOLON       ;                   
31     12     1      BEGIN           begin               
32     13     5      ID              x                   
33     13     7      ASSIGN          :=                  
34     13     10     INT             0                   
35     13     11     SEMICOLON       ;                   
36     14     5      ID              y                   
37     14     7      ASSIGN          :=                  
38     14     10     INT             0                   
39     14     11     SEMICOLON       ;                   
40     15     5      ID              sum                 
41     15     9      ASSIGN          :=                  
42     15     12     INT             0                   
43     15     13     SEMICOLON       ;                   
44     16     5      ID              count               
45     16     11     ASSIGN          :=                  
46     16     14     INT             0                   
47     17     1      END             end                 
48     17     4      SEMICOLON       ;                   
49     19     1      PROCEDURE       procedure           
50     19     11     ID              calculate           
51     19     20     SEMICOLON       ;                   
52     20     1      VAR             var                 
53     20     5      ID              temp                
54     20     9      SEMICOLON       ;                   
55     21     1      BEGIN           begin               
56     22     5      ID              temp                
57     22     10     ASSIGN          :=                  
58     22     13     ID              x                   
59     22     15     PLUS            +                   
60     22     17     ID              y                   
61     22     18     SEMICOLON       ;                   
62     23     5      ID              sum                 
63     23     9      ASSIGN          :=                  
64     23     12     ID              sum                 
65     23     16     PLUS            +                   
66     23     18     ID              temp                
67     23     22     SEMICOLON       ;                   
68     24     5      ID              count               
69     24     11     ASSIGN          :=                  
70     24     14     ID              count               
71     24     20     PLUS            +                   
72     24     22     INT             1                   
73     25     1      END             end                 
74     25     4      SEMICOLON       ;                   
75     27     1      BEGIN           begin               
76     28     5      CALL            call                
77     28     10     ID              init                
78     28     14     SEMICOLON       ;                   
79     31     5      ID              x                   
80     31     7      ASSIGN          :=                  
81     31     10     ID              min                 
82     31     13     SEMICOLON       ;                   
83     32     5      WHILE           while               
84     32     11     ID              x                   
85     32     13     LT              <                   
86     32     15     ID              max                 
87     32     19     DO              do                  
88     33     5      BEGIN           begin               
89     34     9      IF              if                  
90     34     12     ODD             odd                 
91     34     16     ID              x                   
92     34     18     THEN            then                
93     35     9      BEGIN           begin               
94     36     13     ID              sum                 
95     36     17     ASSIGN          :=                  
96     36     20     ID              sum                 
97     36     24     PLUS            +                   
98     36     26     ID              x                   
99     36     27     SEMICOLON       ;                   
100    37     13     WRITE           write               
101    37     18     LPAREN          (                   
102    37     19     ID              sum                 
103    37     22     RPAREN          )                   
104    38     9      END             end                 
105    38     12     SEMICOLON       ;                   
106    39     9      ID              x                   
107    39     11     ASSIGN          :=                  
108    39     14     ID              x                   
109    39     16     PLUS            +                   
110    39     18     ID              step                
111    40     5      END             end                 
112    40     8      SEMICOLON       ;                   
113    43     5      READ            read                
114    43     9      LPAREN          (                   
115    43     10     ID              y                   
116    43     11     COMMA           ,                   
117    43     13     ID              z                   
118    43     14     RPAREN          )                   
119    43     15     SEMICOLON       ;                   
120    46     5      IF              if                  
121    46     8      ID              y                   
122    46     10     GT              >                   
123    46     12     ID              z                   
124    46     14     THEN            then                
125    47     5      BEGIN           begin               
126    48     9      ID              x                   
127    48     11     ASSIGN          :=                  
128    48     14     ID              y                   
129    48     15     SEMICOLON       ;                   
130    49     9      ID              y                   
131    49     11     ASSIGN          :=                  
132    49     14     ID              z                   
133    49     15     SEMICOLON       ;                   
134    50     9      CALL            call                
135    50     14     ID              calculate           
136    51     5      END             end                 
137    52     5      ELSE            else                
138    53     5      BEGIN           begin               
139    54     9      ID              x                   
140    54     11     ASSIGN          :=                  
141    54     14     ID              z                   
142    54     15     SEMICOLON       ;                   
143    55     9      CALL            call                
144    55     14     ID              calculate           
145    56     5      END             end                 
146    56     8      SEMICOLON       ;                   
147    59     5      WRITE           write               
148    59     10     LPAREN          (                   
149    59     11     ID              sum                 
150    59     14     COMMA           ,                   
151    59     16     ID              count               
152    59     21     RPAREN          )                   
153    60     1      END             end                 

================================================================================

Lexical analysis completed, no errors.

================================================================================
